src/main/java/net/iponweb/disthene/service/aggregate/AggregateService.java

```diff
@@ -0,0 +1,179 @@
+    }
+
+    private void aggregate(Metric metric) {
+        ConcurrentMap<MetricKey, AverageRecord> timestampMap = getTimestampMap(metric.getTimestamp());
+        MetricKey destinationMetricKey = new MetricKey(
+                metric.getTenant(), metric.getPath(),
+                metric.getRollup(), metric.getPeriod(),
+                metric.getTimestamp());
+        getAverageRecord(timestampMap, destinationMetricKey).addValue(metric.getValue());
+    }
+
+    private void flush() {
+        Collection<Metric> metricsToFlush = new ArrayList<>();
+
+        while(accumulator.size() > 0 && (accumulator.firstKey() < DateTime.now(DateTimeZone.UTC).getMillis() / 1000 - distheneConfiguration.getCarbon().getAggregatorDelay())) {
+            logger.debug("Adding base rollup flush for time: " + (new DateTime(accumulator.firstKey() * 1000)) + " (current time is " + DateTime.now(DateTimeZone.UTC) + ")");
+
+            // Get the earliest map
+            ConcurrentMap<MetricKey, AverageRecord> timestampMap = accumulator.pollFirstEntry().getValue();
+
+            for(Map.Entry<MetricKey, AverageRecord> entry : timestampMap.entrySet()) {
+                metricsToFlush.add(new Metric(entry.getKey(), entry.getValue().getAverage()));
+            }
+        }
+
+        if (metricsToFlush.size() > 0) {
+            doFlush(metricsToFlush, getFlushRateLimiter(metricsToFlush.size()));
+        }
+    }
+
+    private RateLimiter getFlushRateLimiter(int currentBatch) {
+        /*
+        The idea is that we'd like to be able to process ALL the contents of accumulator in 1/2 time till next rollup time.
+        Doing so, we hope to never limit as much as to saturate the accumulator and to heavily fall back
+         */
+
+        // 100 is an arbitrary small number here
+        double rate = Math.max(100, 2 * currentBatch / baseRollup.getRollup());
+
+        return RateLimiter.create(rate);
+
+    }
+
+    private void doFlush(Collection<Metric> metricsToFlush, RateLimiter rateLimiter) {
+        // We'd like to feed metrics in a more gentle manner here but not allowing the queue to grow.
+
+        logger.debug("Flushing rollup metrics (" + metricsToFlush.size() + ")");
+        if (rateLimiter != null) {
+            logger.debug("QPS is limited to " + (long) rateLimiter.getRate());
+        }
+
+        for(Metric metric : metricsToFlush) {
+            if (!shuttingDown && rateLimiter != null) {
+                rateLimiter.acquire();
+            }
+            bus.post(new MetricStoreEvent(metric)).now();
+        }
+    }
+
+    public synchronized void shutdown() {
+        // disable rate limiters
+        shuttingDown = true;
+
+        scheduler.shutdown();
+
+        Collection<Metric> metricsToFlush = new ArrayList<>();
+
+        for(Map.Entry<Long, ConcurrentMap<MetricKey, AverageRecord>> entry : accumulator.entrySet()) {
+            for(Map.Entry<MetricKey, AverageRecord> innerEntry : entry.getValue().entrySet()) {
```

Comment:
it's java8? In java8 you can use foreach (https://www.mkyong.com/java8/java-8-foreach-examples/) which make syntax simplier

Reply:
"First of all, it's strongly recommended to run it with Java 8. Even though this software is fully compatible with Java 7."
Let's leave it how it is now.

Reply:
great!